### PR TITLE
fix: continue on errors with missing protected branches

### PR DESCRIPTION
**Pull request by vgropp**

### CODE REVIEW

File: pkg/gitlab/project_manager.go [hunk 1]
```diff
@@ -86,10 +86,13 @@ func (m *ProjectManager) EnsureBranchesAndProtection(project gitlab.Project, dry
 		return err
 	}
 
+	branchesWithError := []string{}
```
Comment (PirminTapken) [inline pkg/gitlab/project_manager.go:89]:
```suggestion
	branchesWithError := []error{}
```

File: pkg/gitlab/project_manager.go [hunk 2]
```diff
@@ -107,7 +110,8 @@ func (m *ProjectManager) EnsureBranchesAndProtection(project gitlab.Project, dry
 		// Remove protections (if present)
 		if resp, err := m.protectedBranchesClient.UnprotectRepositoryBranches(project.ID, b.Name); err != nil &&
 			(resp == nil || resp.StatusCode != http.StatusNotFound) {
-			return fmt.Errorf("failed to unprotect branch %v before protection: %v", b.Name, err)
+			branchesWithError = append(branchesWithError, fmt.Sprintf("%s: %v", b.Name, err))
```
Comment (PirminTapken) [inline pkg/gitlab/project_manager.go:113]:
```suggestion
			branchesWithError = append(branchesWithError, fmt.Errorf("%s: %w", b.Name, err)
```

`%w` wraps the error, so you can get it back later by unwrapping the result from `fmt.Errorf`

File: pkg/gitlab/project_manager.go [hunk 3]
```diff
@@ -118,10 +122,15 @@ func (m *ProjectManager) EnsureBranchesAndProtection(project gitlab.Project, dry
 
 		// (Re)add protections
 		if _, _, err := m.protectedBranchesClient.ProtectRepositoryBranches(project.ID, opt); err != nil {
-			return fmt.Errorf("failed to protect branch %s: %v", b.Name, err)
+			branchesWithError = append(branchesWithError, fmt.Sprintf("%s: %v", b.Name, err))
```
Comment (PirminTapken) [inline pkg/gitlab/project_manager.go:125]:
```suggestion
			branchesWithError = append(branchesWithError, fmt.Errorf("%s: %w", b.Name, err)
```

File: pkg/gitlab/project_manager.go [hunk 4]
```diff
@@ -118,10 +122,15 @@ func (m *ProjectManager) EnsureBranchesAndProtection(project gitlab.Project, dry
...
+			continue
 		}
 	}
 
+	if len(branchesWithError) > 0 {
+		return fmt.Errorf("failed to protect branches: %s", strings.Join(branchesWithError, ","))
```
Comment (PirminTapken) [inline pkg/gitlab/project_manager.go:131]:
```suggestion
		return fmt.Errorf("failed to protect branches: %s", func() string {
		        l := make([]string, len(branchesWithError))
		        for i, e := range branchesWithError {
		                l[i] = e.Error()
		        }
		        return strings.Join(l, ",")
		}())
```

Comment (PirminTapken) [reply pkg/gitlab/project_manager.go:131]: Using the inline function here is horrible. My intention with this whole change to keep errors being errors as long as possible. The alternative would be to declare `branchesWithError` like this:

```
type errorList []error

func (l errorList) Error() string {
        sl := make([]string, len(l))
        for i, e := range l {
                sl[i] = e.Error()
        }
        return strings.Join(sl, ",")
}

branchesWithError := errorList{}

// Then later in line 131

return fmt.Errorf("failed to protect branches: %w", branchesWithError)
```
And the the receiving function could unwrap the error list again

This declaration should even work inside the function body

Comment (vgropp) [reply pkg/gitlab/project_manager.go]: i used this suggestion

File: cmd/sync.go [hunk 5]
```diff
@@ -38,8 +38,13 @@ var syncCmd = &cobra.Command{
 		for index, project := range projects {
 			logger.Infof("Processing project #%d: %s", index+1, project.PathWithNamespace)
 
+			if err := manager.EnsureDefaultBranch(project, env.Dryrun); err != nil {
```
Comment (PirminTapken) [inline cmd/sync.go:41]: <3

File: pkg/gitlab/project_manager.go [hunk 6]
```diff
@@ -78,18 +78,24 @@ func (m *ProjectManager) ComplianceReady() bool {
...
+
+func (m *ProjectManager) EnsureBranchProtection(project gitlab.Project, dryrun bool) error {
+	branchesWithError := errorList{}
 
 	for _, b := range m.config.ProtectedBranches {
 		protectedBranch, _, err := m.protectedBranchesClient.GetProtectedBranch(project.ID, b.Name)
 		if err != nil {
 			m.logger.Warnf("failed to get protected branch %v: %v", b.Name, err)
+			continue
 		} else {
```
Comment (PirminTapken) [inline pkg/gitlab/project_manager.go:99]: The `continue` allows refactoring of the `else` case into the normal code flow

**Pull request merged**